### ISSUE TITLE
Trailing slash resolves in 404

### ISSUE THREAD

**Issue by preetamslot** [vite-pwa/astro#23]
Hi,
The pwa works great but I would like to know how to redirect
`/blog/post/` to `blog/post` trailing slashes in the url are not resolving.

It is a static build

```
build: {
    format: "file"
  },
```

```
AstroPWA({
      mode: "production",
      base: "/",
      scope: "/",
      includeAssets: [`icons/${process.env.THEME}/favicon.svg`],
      registerType: "autoUpdate",
      manifest: manifest,
      workbox: {
        navigateFallback: "/404",
        globPatterns: ["**/*.{js,css,html,txt,json,svg,woff,woff2,ico,xml}"],
      },
      devOptions: {
        enabled: false,
        navigateFallbackAllowlist: [/^\/404$/],
      },
    }),
```

**Comment by felixroos** [comment]
I have a similar problem, where trailing slashes always end up on the root path

**Comment by userquin** [comment]
Can you configure tralingSlash in Astro? https://github.com/vite-pwa/astro/blob/main/src/index.ts#L54

**Comment by felixroos** [comment]
> Can you configure tralingSlash in Astro? https://github.com/vite-pwa/astro/blob/main/src/index.ts#L54

thanks for the quick answer. yes I've now done that, trying both combinations. 

- `trailingSlash: 'always'` + `build.format: 'directory'`: only trailing slashes work, without trailing slashes the service worker loads the root page. A hard refresh gives a 404
- `trailingSlash: 'never'` + `build.format: 'file'`: same as above but flipped.. this is how I have currently deployed it.

Example Links: 

- <https://strudel.tidalcycles.org/workshop/getting-started> correct page is loaded
- <https://strudel.tidalcycles.org/workshop/getting-started/> root route is loaded by the SW. cache refresh gives 404

The optimal case would be if it did not matter if a trailing slash is used or not.
I am not entirely sure if this is a vite / astro / vite-pwa or github pages problem...

**Comment by felixroos** [comment]
Just found: https://github.com/slorber/trailing-slash-guide which explains the [behavior of github pages for different links](https://slorber.github.io/trailing-slash-guide/), notably:

[/file](https://slorber.github.io/trailing-slash-guide/file) - [✅](https://slorber.github.io/trailing-slash-guide/trailing-slash-guide/file)
[/file/](https://slorber.github.io/trailing-slash-guide/file/) - [💢 404](https://slorber.github.io/trailing-slash-guide/trailing-slash-guide/file/)
[/folder](https://slorber.github.io/trailing-slash-guide/folder) - [➡️ /folder/](https://slorber.github.io/trailing-slash-guide/trailing-slash-guide/folder)
[/folder/](https://slorber.github.io/trailing-slash-guide/folder/) - [✅](https://slorber.github.io/trailing-slash-guide/trailing-slash-guide/folder/)

This further confirms that `build.format: 'file'` will cause a 404 with a trailing slash.
Without using a service worker, it looks like it would work with `build.format: 'directory'`, as it will just do a redirect.
So is suspect that the service worker will interpret urls without trailing slashes as a file and for some reason serve the root file.

**Comment by felixroos** [comment]
I have done a quick test repro with a simple astro project that builds using `build.format: 'directory'`: https://github.com/felixroos/astro-test/

Both of these urls work:

- https://felixroos.github.io/astro-test/test/
- https://felixroos.github.io/astro-test/test (redirects to above url)

So it seems the problem is really service worker related..

**Comment by userquin** [comment]
IIRC there is a directoryIndex and cleanUrl in workbox option:
https://developer.chrome.com/docs/workbox/modules/workbox-precaching/#directory-index

**Comment by userquin** [comment]
I Will try to check your repro tmr.

**Comment by felixroos** [comment]
> IIRC there is a directoryIndex and cleanUrl in workbox option: https://developer.chrome.com/docs/workbox/modules/workbox-precaching/#directory-index

> EDIT: maybe both options wth another names, the link is about using the workbox modules.

On first sight that sounds like it would help in this case!

> I Will try to check your repro tmr.

Thanks!

For completeness sake, I've now done a [full reproduction](https://github.com/felixroos/astro-test-pwa) that shows how adding @vite-pwa/astro to the mix results in wrong caching behavior: 

- https://felixroos.github.io/astro-test-pwa/ ✅ works
- https://felixroos.github.io/astro-test-pwa/test ✅ works, redirects to test/ on hard refresh (!)
- https://felixroos.github.io/astro-test-pwa/test/ ❌ loads index.astro, works without hard refresh (once)

**Comment by userquin** [comment]
@felixroos 

can you deploy using `mode: 'development'` in pwa options (https://github.com/vite-pwa/astro/blob/main/examples/pwa-simple/astro.config.mjs#L14)? this way you will see the sw logs, I'll change the logic in the pwa integration to allow configure external `manifestTransform`, rn the pwa integration will add always a new one to handle Astro directory and trailing slash options.

**Comment by userquin** [comment]
Since you're deploying to gh pages, the scope is the project name without final slash (check the sw pre cache manifest in your deployed app) and so the sw maybe intercepting the `/test/`, I'll fix that also (GH Pages redirecting to `/astro-test-pwa/` when requesting `/astro-test-pwa`

**Comment by userquin** [comment]
Preparing PR to not modify the sw precache manifest when using directory, on local works for `/test/`, `/test, `/test/about` and `/test/about/`, check the pwa-simple example (added same folders).

**Comment by userquin** [comment]
Astro build is inconsistent, why `404/index.html` missing? only 404.html in root dist folder

**Comment by userquin** [comment]
@felixroos can you test in local with current PR or review it (GH Pages maybe will fail from your url sinde `/test/` will return 404)?

**Comment by userquin** [comment]
You can open the pwa-example using [this StackBlitz repro](https://stackblitz.com/fork/github/vite-pwa/astro/tree/userquin/fix-allow-custom-manifest-transform/examples/pwa-simple)

**Comment by felixroos** [comment]
> can you deploy using `mode: 'development'` in pwa options? this way you will see the sw logs

[deployed](https://felixroos.github.io/astro-test-pwa/), but it only logs "SW registered: ..." atm

> @felixroos can you test in local with current PR or review it 

can test tomorrow (if it's still relevant)

> (GH Pages maybe will fail from your url since /test/ will return 404)?

with `build.format: 'directory'`, `/test/` should work, as it will lead to `/test/index.html` which exists.
it also works rn: https://felixroos.github.io/astro-test-pwa/test/ (with hard refresh)

the stackblitz example seems to work fine!

**Comment by userquin** [comment]
@felixroos don't forget to include `registerType: 'autoUpdate'` in your pwa options, the new sw will be awaiting to confirm update (default is register type is `prompt`).

It seems SB has fixed the `Open in new window` link.

I'll check why `mode` is not working, maybe Astro build changing it, on local workbox messages are there:
![imagen](https://github.com/vite-pwa/astro/assets/6311119/c03805d9-6d09-47a2-8e2b-ce6606b92b5d)

**Comment by felixroos** [comment]
https://felixroos.github.io/astro-test-pwa/ now deployed with autoUpdate + menu

**Comment by userquin** [comment]
@felixroos can you try building the branch and using file protocol to install it in your repo?
- clone this repo (all branches)
- checkout `userquin/fix-allow-custom-manifest-transform`  branch
- `pnpm install && pnpm build && pnpm pack`
- copy tgz file and paste in your repo (add it to git)
- in your repo: `pnpm remove -D @vite-pwa/astro && pnpm add -D ./vite-pwa-astro-0.1.3.tgz`

**Comment by felixroos** [comment]
done: https://felixroos.github.io/astro-test-pwa/ works!

**Comment by userquin** [comment]
@felixroos Now the dilemma in this repo with the PR, I'm afraid to put it as is, perhaps with a flag to activate it?

**Comment by felixroos** [comment]
> @felixroos Now the dilemma in this repo with the PR, I'm afraid to put it as is, perhaps with a flag to activate it?
> 
> EDIT: in any case, the owner can plug its own `manifestTransform` and do whatever (s)he wants

You mean because an opt-in flag won't break existing setups that rely on the old behavior?

**Comment by userquin** [comment]
> You mean because an opt-in flag won't break existing setups that rely on the old behavior?

YES

**Comment by felixroos** [comment]
I'm not sure as I have too little knowledge of this project to give my opinion on it. Of course, having less config options is always a plus, but maybe the old behavior is sometimes desired, not sure. From my perspective this looks more like a bug that should probably be fixed by default, but maybe it introduces some regression.
Maybe start with the feature flag and test it with a handful of existing projects to see if it has downsides?
Either way, I would be very happy to be able to use this as it will finally fix this annoying bug I was having for some time now.

**Comment by felixroos** [comment]
> > You mean because an opt-in flag won't break existing setups that rely on the old behavior?
> 
> YES (I'll check it next week, maybe you can update the PR, I need to rest...)

I can see what I can do, please give yourself some rest. I can also wait or use the tgz

**Comment by userquin** [comment]
The problem, I don't use Astro, I switched to Nuxt 3... This repo should be in Astro repo, but the PR was closed a few times without any explanation and finally added here (íles integration for example in the repo not in Vite Org)

**Comment by userquin** [comment]
@felixroos I'll add it under experimental, like PWA VitePress Integration, check this PR: https://github.com/vite-pwa/vitepress/pull/23

**Comment by felixroos** [comment]
> @felixroos I'll add it under experimental, like PWA VitePress Integration, check this PR: [vite-pwa/vitepress#23](https://github.com/vite-pwa/vitepress/pull/23)

thanks!

**Comment by userquin** [comment]
@felixroos can you run the same process with the branch in this PR in you GH repro and deploy to GH Pages?
- clone this repo (all branches)
- checkout `userquin/feat-experimental-trailing-slashes-handler`  branch
- `pnpm install && pnpm build && pnpm pack`
- **in your repo**: run `pnpm remove -D @vite-pwa/astro` then remove `vite-pwa-astro-0.1.3.tgz` file
- copy new tgz file and paste it in your repo (add it again to git)
- in your repo: `pnpm add -D ./vite-pwa-astro-0.1.3.tgz`
- include in pwa options:
```ts
experimental: { directoryAndTrailingSlashHandler: true }
``` 
- in your repo `pnpm build && pnpm preview` and check it is working
- deploy to GH Pages

I'm preparing the docs, the experimental option added in VitePress cannot be added, Astro will clone the PWA options and cannot be referenced, it will require a change in `vite-plugin-pwa`.

**Comment by userquin** [comment]
If you cannot checkout the new branch, just clone the new branch directly

**Comment by userquin** [comment]
@felixroos released v0.1.4

**Comment by felixroos** [comment]
@userquin thanks for the release! I am now using the new flag and it works :)
